src/chess/bitboard.cc

```diff
@@ -275,46 +275,47 @@ const int kQueenCastleIndex =
 
 Move::Move(const std::string& str, bool black) {
   if (str.size() < 4) throw Exception("Bad move: " + str);
-  from_ = BoardSquare(str.substr(0, 2), black);
-  to_ = BoardSquare(str.substr(2, 2), black);
+  SetFrom(BoardSquare(str.substr(0, 2), black));
+  SetTo(BoardSquare(str.substr(2, 2), black));
   if (str.size() != 4) {
     if (str.size() != 5) throw Exception("Bad move: " + str);
     switch (str[4]) {
       case 'q':
       case 'Q':
-        promotion_ = Promotion::Queen;
+        SetPromotion(Promotion::Queen);
         break;
       case 'r':
       case 'R':
-        promotion_ = Promotion::Rook;
+        SetPromotion(Promotion::Rook);
         break;
       case 'b':
       case 'B':
-        promotion_ = Promotion::Bishop;
+        SetPromotion(Promotion::Bishop);
         break;
       case 'n':
       case 'N':
-        promotion_ = Promotion::Knight;
+        SetPromotion(Promotion::Knight);
         break;
       default:
         throw Exception("Bad move: " + str);
     }
   }
 }
 
-uint16_t Move::as_packed_int() const {
-  if (promotion_ == Promotion::Knight) {
-    return from_.as_int() * 64 + to_.as_int();
-  } else {
-    return static_cast<int>(promotion_) * 64 * 64 + from_.as_int() * 64 +
-           to_.as_int();
-  }
-}
+    uint16_t Move::as_packed_int() const {
```

Comment:
Given the fuss you're going to to maintain just such a packed int, I'd say this can be simplified to just `return data_`, since this function is considered as a black box from the outside.

Reply:
I'd planned on that, but the current code is not equivalent to that. If it ignored the promotion when it was Promotion::None, that'd be equivalent to just masking off the castling bit, but as it is it ignores knight promotions, and adding that logic with the packed int wouldn't be much simpler than what's already there.

Reply:
I've pinged crem asking if it's supposed to be ignoring knight promotions, if that's unintended behaviour then I agree that it'd be better to just return data.

Reply:
Done